Melt parameterisation

dngoldberg opened the pull request:
Enhancement

This pull request makes changes to config.py, solver.py, inout.py and model.py

PURPOSE
----------
All ice shelves have a basal mass balance (melt/freeze) and this is very dependent on the draft of the ice shelf. As using an ocean model is not feasible many studies employ a very simple depth-dependent parameterisation of melt, and so in doing this fenics_ice is consistent with many other ice sheet modelling studies. 

The paramerisation i've chosen is based on representative profiles of temperature in the amundsen sea, which is near freezing at the surface, and then increases across a thermocline to the temperature of circumpolar deep water. The parameterisation i have chosen is

m = 0.5 M_{max} * ( 1.0 + tanh((D - M_d/2.0) / (M_d/4.0) ) )

where m is meltrate (in meters per year), D is depth (positive where below SL) and M_d and M_{max} are parameters. The aim is to choose a parameterisation that uses as few parameters as possible. I am thinking ahead to inversion with a time-dependent model allowing us to invert for melting parameters as well as alpha/beta, and i do not want to introduce new spatially varying controls.

IMPLEMENTATION
---------------------

def_thickadv_eq() is modified to define the object "bmelt" in terms of thickness. Additionally bmelt is projected as-needed for diagnostics. 

CONDITIONALS: Several ufl conditionals were created in solver.py

1. melt_conditional: Melt should not occur under ice that is grounded. Based on others' research, it is best to only allow melting in cells that satisfy the floatation condition in a cell-average sense, that is that the average weight of the ice column is less than the average weight of the ocean column (i.e. depth of bed). Further melt can be limited through setting the field "melt_domains" to zero (see below). To ensure that this criterion is based on cell-average values I created new functions in a DG(0) space (including thickness, in case we define thickness space to be other than DG(0)).

2. melt_depth_condtional, and melt_max_conditional: Because of heterogeneities in bathymetry different ice shelves melt at different rates. We should be able to represent this by allowing M_d and M_{max} to vary per ice shelf. A field "melt_domains" is initialised from a file (if no file, it is uniformly equal to 1.0). This should define the regions where different (M_d, M_{max)) pairs are used.

**Unfortunately I do not know how to allow for more than 2 sets of parameters.** This would be nice to be able to do.

The use of project() and localprojectionsolver() should be checked to make sure it is not hurting performance

DIAGNOSTICS / OUTPUT:
----------------------------
timestep() is changed so as to limit time-dependent output. This is not completely related to melt but wrapped into this pull request. Output is now through write_variable(), and it is written at predefined intervals based on a parameter write_frequency(). Because of the use of write_variable() there are separate files for each time level written. write_variable() also now allows for h5 output. 

CONFIGURATION:
--------------------

The .toml file now has new parameters

melt_domains_data_file = ""
melt_domains_field_name = ""

in the "io" section, and 

save_frequency: float = 0

in the "time" section, and a new "melt" section with parameters

use_melt_parameterisation: bool = False
depth_therm_domain_1: float = 0.0
depth_therm_domain_2: float = 0.0
max_melt_domain_1: float = 0.0
max_melt_domain_2: float = 0.0

There is no breaking change -- and the melt parameterisation cannot be tested through the existing pytest suite as ice_stream has no floating ice. An issue can be raised to create a smith pytest.

bearecinos commented:
Thanks @dngoldberg I will review this and try to merge between today and tomorrow... I am almost done with my changes I think I will merge your PR first and then adapt my version...
Probably then its good to discuss then tomorrow how can we test this on smith

bearecinos commented:
is it fare to say the toml should look like this:

```
[io]

melt_domains_data_file = ""
melt_domains_field_name = ""

[time]
save_frequency: float = 0

[melt]

use_melt_parameterisation: bool = False
depth_therm_domain_1: float = 0.0
depth_therm_domain_2: float = 0.0
max_melt_domain_1: float = 0.0
max_melt_domain_2: float = 0.0
```

dngoldberg commented:
hi @bearecinos thanks -- i did some questionable things though (lots of extra functions, lots of projections, and lots of use of conditionals) and i'd want to make sure @jrmaddison is OK with them in case they hurt performance..

dngoldberg commented:
once merged though i'd like to discuss what is high priority next -- (1) pytest for this? (2)[ issue 67](https://github.com/EdiGlacUQ/fenics_ice/issues/67)?  (3) sliding law addition?

dngoldberg commented:
> is it fare to say the toml should look like this:
> 
> ```
> [io]
> 
> melt_domains_data_file = ""
> melt_domains_field_name = ""
> 
> [time]
> save_frequency: float = 0
> 
> [melt]
> 
> use_melt_parameterisation: bool = False
> depth_therm_domain_1: float = 0.0
> depth_therm_domain_2: float = 0.0
> max_melt_domain_1: float = 0.0
> max_melt_domain_2: float = 0.0
> ```

yes! though with other params under "time" and "io".

jrmaddison commented on fenics_ice/solver.py:
```diff
@@ -1077,6 +1106,50 @@ def bglen_data_conditional(self, H, glen_mask, H_float=None):
+        depth2 = constants.depth_therm_domain_2
+
+        # Note: cell=triangle just suppresses a UFL warning ("missing cell")
+        melt_depth = ufl.operators.Conditional(ufl.eq(self.model.melt_domains,1.0),
+                                          Constant(depth1, cell=triangle, name="Depth Domain 1"),
+                                          Constant(depth2, cell=triangle, name="Depth Domain 2"))
```
Could use `(self.model.melt_domains - 1) * depth2 - (self.model.melt_domains - 2) * depth1` to avoid the conditional -- although might then want to add a check that `self.model.melt_domains` equals 1 or 2 everywhere

dngoldberg replied:
i think this is resolved.. got rid of these conditionals

jrmaddison reviewed (commented):
Subdomains are an alternative, but probably not any easier. It might be easier to mark more domains by setting up DG0 parameter fields first -- but I expect only if a larger number of regions are needed.

`project`s use default parameters -- but we should probably look at these separately.

dngoldberg commented:
hi @jrmaddison @bearecinos i have modified this PR slightly -- 

1) i changed the way parameters are set (there are 2 parameters at a location that determine melt rate), now M_{max} and M_d can be imposed globally through toml parameters, or through 2 input .h5 files, which override the parameters in the .toml. I've also made the [melt] section of the toml optional, so that i didn't need to change all example cases!

2) i've merged my changes into main, which now includes bea's phase names. Bea -- to do this i had to change your code to write diagnostics in solver.timestep(), i now use write_variable and pass the phase name and suffix and diags location. I wonder if you can look at these changes in solver?

dngoldberg commented on fenics_ice/solver.py:
```diff
@@ -546,7 +563,7 @@ def def_thickadv_eq(self):
 
             # basal melting
-            + bmelt*Ksi*fl_ex*dx
+            + bmelt*Ksi*dx
```
has been replaced with another conditional tailored to melting

dngoldberg commented on fenics_ice/solver.py:
```diff
@@ -514,6 +516,21 @@ def def_thickadv_eq(self):
+          depth = rhoi/rhow * H_np
+          meltcond = self.melt_conditional(H_np,H_DG)
+          meltdepthparam = self.melt_depth_conditional()
+          meltmaxparam = self.melt_max_conditional()
```
i think this is now done in latest commit..

dngoldberg commented on fenics_ice/solver.py:
```diff
@@ -587,8 +605,13 @@ def timestep(self, adjoint_flag=1, qoi_func=None ):
-        save_every_tstep = config.save_every_tstep
-
+        save_frequency = config.save_frequency
+        n_save_frequency = int(np.round(save_frequency/dt))
```
@jrmaddison  do you mean

int(min(max(np.round(save_frequency/dt+0.5),1),n_steps))  ?

jrmaddison replied:
Yes, I think so

bearecinos reviewed (commented):
Hi @dngoldberg 
I've reviewed solver.py and config.py, things look good to me. As we don't have a test for this yet the best I think its to merge it and re run the smith glacier repo experiments ... 

if you and @jrmaddison are happy with this lets merge it!

After inspecting how this works within the smith glacier framework maybe we can add a small test using smith?

bearecinos commented on fenics_ice/config.py:
```diff
@@ -424,7 +444,7 @@ class IOCfg(ConfigPrinter):
 
     log_level: str = "info"
-    output_var_format: str = "both"
+    output_var_format: str = "pvd"
```
The choice on output its only for diagnostic files or things that you write with ``write_variable`` so the choice of output in the toml does not affect the workflow

dngoldberg replied:
yes of course. what i mean is, would 

"pvd h5 xml"

or 

"h5 xml" 

be admissible as values in the toml file? I know now how to set xml as output, or pvd or all three.. how to set both pvd and xml, but not h5, for instance?

bearecinos replied:
@dngoldberg good point! we also need to add all those files to this list of strings in [config.py#L472-L475](https://github.com/EdiGlacUQ/fenics_ice/blob/723ed2f5a243cf989151a42917f03af2fc14e282/fenics_ice/config.py#L472-L475):
```
 assert self.output_var_format in ["pvd", "xml", "h5", "all"], \
            "Invalid variable output file format"
```
only 'h5' is missing at the moment

bearecinos replied:
yes we can add ``output_var_format`` to be a list later on ... no worries

dngoldberg commented:
thats great bea thanks for looking. 

> After inspecting how this works within the smith glacier framework maybe we can add a small test using smith?

that sounds ideal! i have tested run_forward() already.. as you say nothing to compare to but melt rates look reasonable at the end, and the grounding line retreats a few 10s of km which i'd expect. As for change in VAF im not sure! we can see.

ill merge in the latest from main, make the change to config.py above, and then i will merge into main?

i don't think i have merged a PR before. to do this.. can i simply push the Green button that says "Merge pull request"?

bearecinos commented:
Yes if there are no conflicts you can merge this, at the moment your branch is behind 

`` 7 commits behind EdiGlacUQ/fenics_ice:main.``

Maybe click the fetch and merge button and check that you still don't have conflicts ... if so and all the test pass go ahead and merge it.
